### PR TITLE
seccomp: SECCOMP_RET_USER_NOTIF support

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by brauner** [comment]
//Cc @stgraber, @tych0, @hallyn

**Comment by brauner** [comment]
Please do an initial review. It's rough-shape though.

File: src/lxc/attach.c [hunk 1]
```diff
@@ -853,11 +853,24 @@ static int attach_child_main(struct attach_clone_payload *payload)
 
 	if (init_ctx->container && init_ctx->container->lxc_conf &&
 	    init_ctx->container->lxc_conf->seccomp) {
-		ret = lxc_seccomp_load(init_ctx->container->lxc_conf);
+		struct lxc_conf *conf = init_ctx->container->lxc_conf;
+
+		ret = lxc_seccomp_load(conf);
```
Comment (tych0) [inline src/lxc/attach.c]: Just a heads up that this API has changed, you have to do a subsequent seccomp_fd_get(), it doesn't return directly from seccomp_load() any more.

File: src/lxc/attach.c [hunk 2]
```diff
@@ -1311,6 +1324,24 @@ int lxc_attach(const char *name, const char *lxcpath,
...
 
+#if HAVE_SCMP_FILTER_CTX
+		if (conf->seccomp && conf->has_seccomp_notify) {
+			ret = lxc_abstract_unix_recv_fds(ipc_sockets[0],
+							 &conf->seccomp_notify_fd,
+							 1, NULL, 0);
+			if (ret < 0)
+				goto close_mainloop;
+
+			SYSERROR("Retrieved seccomp fd %d", conf->seccomp_notify_fd);
```
Comment (tych0) [inline src/lxc/attach.c]: INFO?

File: src/lxc/seccomp.c [hunk 3]
```diff
@@ -1243,5 +1271,57 @@ void lxc_seccomp_free(struct lxc_conf *conf)
...
+	struct seccomp_notify_proxy_msg msg;
+
+	if (listener_proxy_fd < 0)
+		return minus_one_set_errno(EINVAL);
+
+	ret = seccomp_notif_receive(fd, req);
+	if (ret)
+		return minus_one_set_errno(-ret);
+
+	if (req->data.nr != SCMP_SYS(mknod))
```
Comment (tych0) [inline src/lxc/seccomp.c]: I guess you probably want to drop this, right?

File: src/lxc/start.c [hunk 4]
```diff
@@ -1278,8 +1292,6 @@ static int do_start(void *data)
 
 	/* Setup the container, ip, names, utsname, ... */
 	ret = lxc_setup(handler);
-	close(handler->data_sock[1]);
-	close(handler->data_sock[0]);
```
Comment (tych0) [inline src/lxc/start.c]: These probably still need to be closed on failure paths?

File: src/lxc/attach.c (hunk 1 above)
Comment (brauner) [reply src/lxc/attach.c]: Oh yeah, I'm already doing that though. This is `lxc_seccomp_load()` not `seccomp_load()` and. The former calls `seccomp_fd_get()`. :)

File: src/lxc/attach.c (hunk 2 above)
Comment (brauner) [reply src/lxc/attach.c]: Yip.

File: src/lxc/seccomp.c (hunk 3 above)
Comment (brauner) [reply src/lxc/seccomp.c]: Yes, for sure. That is residual from testing.

File: src/lxc/start.c (hunk 4 above)
Comment (brauner) [reply src/lxc/start.c]: Yeah, also left-over when hacking this in for testing.

File: src/lxc/attach.c (hunk 1 above)
Comment (tych0) [reply src/lxc/attach.c]: Oh, I thought I'd looked for it and didn't see it...

File: src/lxc/attach.c (hunk 2 above)
Comment (brauner) [reply src/lxc/attach.c]: done

File: src/lxc/seccomp.c (hunk 3 above)
Comment (brauner) [reply src/lxc/seccomp.c]: done

File: src/lxc/start.c (hunk 4 above)
Comment (brauner) [reply src/lxc/start.c]: done

File: doc/lxc.container.conf.sgml.in [hunk 5]
```diff
@@ -1873,8 +1873,27 @@ dev/null proc/kcore none bind,relative 0 0
...
       </programlisting>
 
+      <para>
+      Specifying "errno" as action will cause LXC to register a seccomp filter
+      that will cause a specific errno to be returned ot the caller. The errno
+      value can be specified after the "errno" action word.
+      </para>
+
+      <para>
+      Specifying "notif" as action will cause LXC to register a seccomp
```
Comment (tych0) [inline doc/lxc.container.conf.sgml.in]: "notify"

File: src/lxc/lxccontainer.c [hunk 6]
```diff
@@ -5725,3 +5725,23 @@ bool lxc_has_api_extension(const char *extension)
 
 	return false;
 }
+
+int seccomp_notify(struct lxc_container *c, unsigned int cmd, int fd)
```
Comment (tych0) [inline src/lxc/lxccontainer.c]: don't you need to initialize this in lxc_container_new() too?

Comment (brauner) [reply src/lxc/lxccontainer.c]: hahah, yes :D

File: doc/lxc.container.conf.sgml.in (hunk 5 above)
Comment (brauner) [reply doc/lxc.container.conf.sgml.in]: done

File: src/lxc/lxccontainer.c (hunk 6 above)
Comment (brauner) [reply src/lxc/lxccontainer.c]: done

**Comment by brauner** [comment]
jenkins: test this please

File: src/lxc/api_extensions.h [hunk 7]
```diff
@@ -43,6 +43,7 @@ static char *api_extensions[] = {
 	"cgroup_relative",
 	"mount_injection_file",
 	"seccomp_allow_nesting",
+	"seccomp_notify",
```
Comment (tomponline) [inline src/lxc/api_extensions.h]: @brauner is it worth adding an entry to https://github.com/lxc/lxc/blob/master/doc/api-extensions.md for this feature?

Comment (brauner) [reply src/lxc/api_extensions.h]: yeah

**Pull request merged**